gguf-py/gguf/gguf.py

```diff
@@ -546,15 +545,16 @@ def add_val(self, val: Any, vtype: Optional[GGUFValueType] = None, add_vtype: bo
             encoded_val = val.encode("utf8") if isinstance(val, str) else val
             self.kv_data += struct.pack("<Q", len(encoded_val))
             self.kv_data += encoded_val
-        elif vtype == GGUFValueType.ARRAY:
-            ltype = set([GGUFValueType.get_type(item) for item in val])
-            assert len(ltype) == 1, "All items in a GGUF array should be of the same type"
-            self.kv_data += struct.pack("<I", list(ltype)[0])
+        elif vtype == GGUFValueType.ARRAY and isinstance(val, Sequence) and len(val) > 0:
```

Comment:
You can't `add_array` an empty array. This behavior was the same in the original version. It built a `set` of the types and checked that the length was `1`.

I think this approach should be a little nicer. The fact that you can't add empty arrays here seems like it would be confusing from a user perspective but I'm not sure how to get around that.

Reply:
With the set approach you could assert that `len(ltype) < 2`, and instead of `list(ltype)[0]` use `next(iter(ltype), GGUFValueType.UINT8)`. I don't think the element type of an empty array should matter in practice.

Reply:
> With the set approach you could assert that `len(ltype) < 2`

That's true. Dealing with that case isn't an issue with my changed version (which I think is better overall than build a set).

> use [...] `GGUFValueType.UINT8`

Suppose I'm some code loading a GGUF file. Take tokenizer scores for example that would be an array of float. If I'm being careful about checking that types are correct and stuff, it's very likely I'd say "Whoa, I expected a float list but this is a uint8 list!` (Also, I checked and it's a bit scary `llama.cpp` _doesn't_ seem to have checks like that and just happily casts the tokenizer score array to `float *`).

@ggerganov How would you recommend handling this? I think there are three main possibilities:

1. Just don't allow saving empty arrays (current behavior).
2. Use some other arbitrary type (like `uint8`).
3. Add a new type for empty arrays (in that case, type info and length wouldn't need to be stored). Alternatively you could add a new type like `Unit` that could be used for approach \#2.

Reply:
I feel like numpy ndarrays map to this serialization format much better than python lists do. They not only let you make an empty array of a given type, but they let you explicitly ask for float64 or int64, which aren't fully implemented in gguf.py.

Reply:
You have a point. I'm fine with implementing a change where that function can only take numpy arrays but I don't have the authority to make the decision. (Also not sure I understand everything well enough to want to be responsible for making it.)

Reply:
I think that "Just don't allow saving empty arrays (current behavior)." is fine